### PR TITLE
Adding Headers and multiple method outputs

### PR DESCRIPTION
**Pull request by rhysnewell**

### CODE REVIEW

**Comment by wwood** [comment]
Hi Rhys,
A few overall comments:
* Can you add some tests?
* Shouldn't it be `--no-headers` rather than `--headers` ?
* Would you mind indenting after an open function bracket rather than having them flat like below?
```
                            let mut input_stream = coverm::genome::mosdepth_genome_coverage_with_contig_names(
                            &bam_files,
                            &genomes_and_contigs,
                            limit_stream,
```
* The big match statement within genome and the corresponding one in contig mode have a lot of repetition. Would it be possible to have code something like this:
```
let coverage_estimator: MosdepthGenomeCoverageEstimator;
match method {
    "mean" => {
coverage_estimator = MeanGenomeCoverageEstimator::new(min_fraction_covered);
}
}

coverage_estimator.add_to_header(&mut htype);
                        let mut input_stream = coverm::contig::contig_coverage(
                        &bam_files,
                        limit_stream,
                        &mut coverage_estimator,
                        print_zeros,
                        flag_filter);
                        let mut out = update_outputs(&mut output_stream, input_stream);
                        output_stream = out.clone();
```
Hope that makes sense
* I think functional code like `pub fn print_header(htype: &mut HeaderTypes){` might be better as a closure, but maybe there's no difference and that's just my (not strong) preference.
* I have not set a good example, but I think documenting the important functions is probably good practice.

Sorry if some of these things are picky, but keeping things like this in check helps in the long run.

**Comment by wwood** [comment]
Also, one more thing I forgot. I don't think we should simply modify `PileupCountsGenomeCoverageEstimator` to also include stdev. Instead, we should make a dedicated 'metabat' output format. That way it can also match the headers etc while leaving CoverM free to decide its own headers for the more vanilla coverage modes. Make sense?

**Comment by rhysnewell** [comment]
Header output now updated to main branch. Closing this pull request as it is no longer needed.

**Comment by wwood** [comment]
Hey,
I'm not sure if you were intending this to be ready for review but I'm guessing it won't compile as-is because the conflicts haven't been fixed properly e.g. there are lines like this still:
```
<<<<<<< HEAD
```

**Comment by rhysnewell** [comment]
Hey, 
No definitely not ready. I closed the pull request by accident trying to fix the conflicts but it all just went a bit whacky and I think I ended up pushing some bad commits. Just ignore them, I'm fixing them.

**Comment by wwood** [comment]
ok np

File: src/mosdepth_genome_coverage_estimators.rs [hunk 1]
```diff
@@ -222,6 +354,15 @@ impl PileupCountsGenomeCoverageEstimator {
             min_fraction_covered_bases: min_fraction_covered_bases
         }
     }
+    pub fn add_to_header(header_types: &mut HeaderTypes) -> &mut HeaderTypes{
+            let coverage_type = "Pileup Counts".to_string();
+            let index = "Index".to_string();
+            let st_dev = "Standard Deviation".to_string();
```
Comment (wwood) [inline src/mosdepth_genome_coverage_estimators.rs]: I'm a little confused. This histogram struct outputs columns for stoit, genome, coverage and number of bases with that coverage (in that genome in that stoit). So, wouldn't the stdev of each row be 0? Guess I'm missing something?

File: src/main.rs [hunk 2]
```diff
@@ -155,29 +196,140 @@ fn main(){
...
+    
     if min_fraction_covered > 1.0 || min_fraction_covered < 0.0 {
         eprintln!("Minimum fraction covered parameter cannot be < 0 or > 1, found {}", min_fraction_covered);
         process::exit(1)
     }
     let print_zeros = !m.is_present("no-zeros");
     let flag_filter = !m.is_present("no-flag-filter");
     let single_genome = m.is_present("single-genome");
-
+    let headers = !m.is_present("remove-headers");
```
Comment (wwood) [inline src/main.rs]: How about we just keep it simple and always print the headers i.e. remove the remove-headers option? It isn't hard for people to run it through `tail -n+2` if they really don't want headers, yeh?

File: src/main.rs [hunk 3]
```diff
@@ -155,29 +196,140 @@ fn main(){
...
     bam_generators: Vec<T>,
-    m: &clap::ArgMatches) {
+    m: &clap::ArgMatches,
+    output_stream: &mut Vec<OutputStream>,
+    htype: &mut HeaderTypes,
+    method: &str) -> Vec<OutputStream> {
+    
 
-    let method = m.value_of("method").unwrap();
+    let out;
```
Comment (wwood) [inline src/main.rs]: Any chance of a better variable name?

**Comment by rhysnewell** [comment]
Hi Ben,

Yep, I’ll get some tests going and all your comments are valid to me. I’ll make the changes as per your suggestions.

Cheers,

Rhys

**Comment by wwood** [comment]
See 80ceacf

**Pull request closed without merging**